[Sol->Yul] External library calls

cameel opened the pull request:
Part of #8078.
Fixes many bugs in #10235.

**This PR is based on #10229. Please don't merge until that one is merged.**

Adds support for `address(L)` and `L.f()`-style external calls.

Semantic tests require linking to pass - that's why this is based on #10229.

Note that there are apparently no call guards generated for libraries yet so `library_address.sol` does not pass even though the code for the call itself is generated correctly.

cameel committed:
IR codegen: Handle address() with library type argument and external library calls

cameel committed:
Test for calling external library functions with mapping parameters (does not work via Yul yet)

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -2959,3 +2966,20 @@ void IRGeneratorForStatements::setLocation(ASTNode const& _node)
+	solAssert(_libraryType.contractDefinition().isLibrary(), "");
+
+	string escapedFullyQualifiedName = _libraryType.contractDefinition().fullyQualifiedName();
+	boost::replace_all(escapedFullyQualifiedName, "\"", "\\\"");
```
What about backslashes?

cameel replied:
Good catch. I thought I could get away with limiting this to quotes but apparently not. I need full-blown escaping. `\n` won't work either and non-printable characters need escaping.

I found `util::escapeAndQuoteString()`. Is the output guaranteed to be safe for Yul/Solidity strings? I added a semantic test for this and also tested manually and it does seem to work.

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -760,12 +761,18 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
+		if (fromType->category() == Type::Category::TypeType && toActualType->category() == Type::Category::Address)
+			define(_functionCall) << linkerSymbol(dynamic_cast<TypeType const&>(*fromType)) << "\n";
+		else
+			define(_functionCall, *_functionCall.arguments().front());
```
I know this has been like this, but why don't we call the conversion function?

chriseth replied:
Oh it does call the conversion function internally!

cameel replied:
Yeah. The conversion is there.

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -760,12 +761,18 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
+		Type const* toActualType = dynamic_cast<TypeType const*>(toType)->actualType();
+
+		if (fromType->category() == Type::Category::TypeType && toActualType->category() == Type::Category::Address)
+			define(_functionCall) << linkerSymbol(dynamic_cast<TypeType const&>(*fromType)) << "\n";
```
Can you add some more assertions here? I know they are inline `linkerSymbol` but this looks a bit weird - why would any TypeType conversion to address lead to linker Symbol?

chriseth replied:
I'm actually just looking at `YulUtilFunctions::conversionFunction`. The checks can be handled there - why would you need to do it here?

cameel replied:
Doing it in `YulUtilFunctions::conversionFunction()` was my first implementation but then I ran into errors about invalid IR. Same thing you can see in the failing tests after your changes: https://app.circleci.com/pipelines/github/ethereum/solidity/10138/workflows/a6059bc2-7bae-408f-8dfb-6f92cc699373/jobs/499676

```yul
            function fun_g_61(vloc_a_36, vloc_b_38) -> vloc__41 {
                let zero_value_for_type_t_bool_2 := zero_value_for_split_t_bool()
                vloc__41 := zero_value_for_type_t_bool_2

                let expr_48 := convert_t_type$_t_contract$_L_$18_$_to_t_address(expr_47_address)
                let expr_49_address := expr_48
                ...
```
```
Invalid IR generated:
:178:81: Error: Identifier not found.
                let expr_26 := convert_t_type$_t_contract$_L_$18_$_to_t_address(expr_25_address)
                                                                                ^-------------^
:191:81: Error: Identifier not found.
                let expr_48 := convert_t_type$_t_contract$_L_$18_$_to_t_address(expr_47_address)
                                                                                ^-------------^
```

The thing is that this is not really a conversion. A conversion function requires a parameter but there's nothing to give it. Well, I could give it the address but then why call it at all? I could also create a dummy variable set to 0 and ignore the parameter completely inside the function but the current solution of just skipping the conversion seemed cleaner to me.

cameel replied:
> Can you add some more assertions here? I know they are inline `linkerSymbol` but this looks a bit weird - why would any TypeType conversion to address lead to linker Symbol?

How about a comment?

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -1831,9 +1829,16 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
+					if (contractType.contractDefinition().isLibrary(), "")
+						define(IRVariable(_memberAccess).part("address")) << linkerSymbol(contractType) << "\n";
+					else
+						define(IRVariable(_memberAccess).part("address"), _memberAccess.expression());
```
How can this be reachable if it is a delegatecall?

cameel replied:
Right. It's not reachable. Removed.

chriseth commented:
Rebased and added (and reverted) commits. Did not touch your commits, so we have to squash some.

cameel commented:
I pushed fixes for some of the comments. The only thing remaining is https://github.com/ethereum/solidity/pull/10230#discussion_r521513228.

chriseth commented:
I think this is now the right way to do it. An expression of type T should result in IR being generated such that all stack items of T are properly defined as yul variables.

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -2158,6 +2159,10 @@ void IRGeneratorForStatements::endVisit(Identifier const& _identifier)
 	{
 		// no-op
 	}
+	else if (dynamic_cast<ImportDirective const*>(declaration))
```
This is not a duplication in my local checkout. Maybe rebasing will fix it.

chriseth commented:
Not sure what to make of the error message `"c&"o\n" is not found.`

cameel commented:
> I think this is now the right way to do it. An expression of type T should result in IR being generated such that all stack items of T are properly defined as yul variables.

ok. I saw that a lot of stuff is done this way but I didn't know that it was a general rule. I thought that not having the no-op conversion was preferable. In that case the PR should fine now after your addition of the missing variable. I squashed all the fixups and extra commits.

> Not sure what to make of the error message "c&"o\n" is not found.

It comes from my multi-source test for weird library names with special characters (both characters that need to be escaped like `"` and `\` as well as non-printable control characters). `ASTImportTest.sh` and `stopAfterParseTests.sh` are choking on it because they don't properly quote file names and assume that values can be safely passed through shell evaluation. I tried to fix the scripts but pure-shell workaround were either horrible or non-portable to Mac. It would be quick and easy in Python but I didn't want to go that far so I'm giving up for now. I removed the test.

BTW, while working on it I also made a proper fix for #9710 so the script does not bail out broken UTF-8. I submit it later because I need to clean up the error handling around it a bit first.
